base_layer/core/src/validation/block_validators.rs

```diff
@@ -99,6 +122,77 @@ impl FullConsensusValidator {
     pub fn new(rules: ConsensusManager) -> Self {
         Self { rules }
     }
+
+    /// Calculates the achieved and target difficulties at the specified height and compares them.
+    fn check_achieved_and_target_difficulty<B: BlockchainBackend>(
+        &self,
+        db: &B,
+        block_header: &BlockHeader,
+    ) -> Result<(), ValidationError>
+    {
+        let pow_algo = block_header.pow.pow_algo;
+        debug!(
+            target: LOG_TARGET,
+            "check_achieved_and_target_difficulty: pow_algo = {}, height = {}", pow_algo, block_header.height,
+        );
+
+        let target = if block_header.height == 0 {
+            1.into()
+        } else {
+            let height = block_header.height - 1;
+            let mut target_difficulties = self.rules.new_target_difficulty(pow_algo, height);
+            for height in (0..=height).rev() {
+                let header = fetch_header(&*db, height)?;
+                if header.pow.pow_algo == pow_algo {
+                    target_difficulties.add_front(header.timestamp(), header.target_difficulty());
+                    if target_difficulties.is_full() {
+                        break;
+                    }
+                }
+            }
```

Comment:
Is this not duplicated here: pub fn fetch_target_difficulty(

Reply:
Not sure if you are referring to `BlockchainDatabase::fetch_target_difficulty` - if so, the reason I duplicate code is because the validator only takes `B: BlockchainBackend` which does not have fetch_target_difficulty. The validators should be refactored have access to BlockchainDatabase, but then the BlockchainDatabase also needs to be refactored to allow the to perform its work within one read transaction. So this is a stop gap.